### DISCUSSION TITLE
RFC: A unified background workers interface for Django

### DISCUSSION THREAD

**Discussion by RealOrangeOne** [celery/celery#9042] [General]
I've recently submitted a Django Enhancement Proposal ("DEP") for [creating a background workers interface for Django](https://github.com/django/deps/pull/86), similar to the existing interfaces for databases and caches. The objective is to better facilitate developers (both working with Django and library maintainers) moving certain tasks from their web processes to separate background runners. Django currently has no first-party support for background workers, instead requiring developers to integrate with an external library (eg Celery).

With a unified interface, developers will be able to more easily integrate with existing background worker libraries, without needing to re-learn a new API. Some backends will be maintained inside Django (eg some for testing, and one using its existing ORM), however the intention is for additional interfaces to integrate with existing libraries, allowing projects to use the queues they already have. 

More details, including rational and proposed API can be found in the [rendered document](https://github.com/RealOrangeOne/deps/blob/background-workers/draft/0014-background-workers.rst).

I'm opening this discussion seeking input from the Celery community on the proposal's API, ensuring it doesn't hinder Django's ability to extend the API for future feature growth, but also to ensure the chosen API is compatible with a Celery integration, even if said integration wouldn't expose all of Celery's features (right now). 

Feel free to reach out below, on [the related DEP PR](https://github.com/django/deps/pull/86), or directly to the contact details in the DEP description.

**Comment by auvipy** [comment]
thanks for bringing this up! I will definitely take part in the design discussion as I have been waiting for it since the time of https://github.com/django/django/pull/4763, although stale now.

**Reply by auvipy** [reply]
Also checking the https://django-q.readthedocs.io/en/latest/ and celery django integration codes might be helpful?

**Reply by RealOrangeOne** [reply]
From the looks of it, integrating with `@shared_task` ought not to be too difficult. So long as celery can deal with these wrapped functions.

Please do take [the current code](https://github.com/realOrangeOne/django-tasks) and have a play around!
